File: C#/309. Best Time to Buy and Sell Stock with Cooldown.cs
public class Solution
{
    public int MaxProfit(int[] prices)
    {
        int[] sell = new int[prices.Length], nosell = new int[prices.Length], dp = new int[prices.Length];
        dp[0] = sell[0] = prices[0];
        int minBuy1 = prices[0], minBuy2 = 0;
        for (int i = 1; i < prices.Length; ++i)
        {
            dp[i - 1] = Math.Max(sell[i - 1], nosell[i - 1]);
            sell[i] = prices[i] > minBuy1 ? nosell[i - 1] + prices[i] - minBuy1 : nosell[i - 1];
            if (sell[i - 1] > nosell[i - 1])
            {
                nosell[i] = sell[i - 1];
                minBuy1 = minBuy2;
            }
            else
                nosell[i] = nosell[i - 1];
            if (minBuy1 > prices[i])
                minBuy1 = prices[i];
        }
        return Math.Max(sell[prices.Length - 1], nosell[prices.Length - 1]);
    }
}

File: python/309. Best Time to Buy and Sell Stock with Cooldown.py
# -*- coding: utf-8 -*-

"""
@Author: xiezizhe
@Date: 24/6/2020 下午11:00
"""


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) <= 1:
            return 0

        sell, buy = [0 for _ in range(len(prices))], [0 for _ in range(len(prices))]
        buy[0] = -prices[0]
        buy_max = buy[0]
        sell_max = 0
        for i in range(1, len(prices)):
            buy[i] = sell_max - prices[i]
            sell[i] = max(buy_max + prices[i], sell[i - 2])

            buy_max = max(buy[i], buy_max)
            sell_max = max(sell[i - 1], sell_max)

        # print(sell)
        return max(sell[-1], sell[-2])
